src/commands/force/user/permsetlicense/assign.ts

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import * as os from 'os';
+import { flags, FlagsConfig, SfdxCommand } from '@salesforce/command';
+import { Aliases, Connection, Messages, User, AuthInfo, Org, UserFields, SfdxError } from '@salesforce/core';
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/plugin-user', 'permsetlicense.assign');
+
+type SuccessMsg = {
+  name: string;
+  value: string;
+};
+
+type FailureMsg = {
+  name: string;
+  message: string;
+};
+
+type Result = {
+  successes: SuccessMsg[];
+  failures: FailureMsg[];
+};
+
+interface PermissionSetLicense {
+  Id: string;
+}
+
+export class UserPermsetLicenseAssignCommand extends SfdxCommand {
+  public static readonly description = messages.getMessage('description');
+  public static readonly examples = messages.getMessage('examples').split(os.EOL);
+  public static readonly requiresUsername = true;
+  public static readonly flagsConfig: FlagsConfig = {
+    name: flags.string({
+      char: 'n',
+      description: messages.getMessage('flags.name'),
+      required: true,
+    }),
+    onbehalfof: flags.array({
+      char: 'o',
+      description: messages.getMessage('flags.onBehalfOf'),
+    }),
+  };
+  private usernames: string[];
+  private readonly successes: SuccessMsg[] = [];
+  private readonly failures: FailureMsg[] = [];
+
+  public async run(): Promise<Result> {
+    try {
+      this.usernames = this.flags.onbehalfof ?? [this.org.getUsername()];
+
+      for (const username of this.usernames) {
+        // Convert any aliases to usernames
+        const aliasOrUsername = (await Aliases.fetch(username)) || username;
+        const connection: Connection = await Connection.create({
+          authInfo: await AuthInfo.create({ username }),
+        });
+        const org = await Org.create({ connection });
+        const user: User = await User.create({ org });
+        const fields: UserFields = await user.retrieve(username);
+
+        const pslName = this.flags.name as string;
+
+        const psl = (
+          await connection.query(
+            `select Id from PermissionSetLicense where DeveloperName = '${pslName}' or MasterLabel = '${pslName}'`
+          )
+        ).records[0] as PermissionSetLicense;
```

Comment:
Will accessing records[0] cause an out of bounds error?

Reply:
yes, consistent with how this is done in `permset:assign`.  

I've used `singleRecordQuery` to handle this pattern in the past, but didn't want to have non-dev-dependency.  https://github.com/mshanemc/plugin-helpers/blob/master/src/queries.ts

We could add something similar to `kit` or just inline it here.

I'm always uncertain how much to "fix" this kind of thing.

Reply:
Great topic for next week, but still leaves the user with a poor experience having to translate the JS out of bounds error to the soql query returning no records.

maybe change to

```
.records.find((r) => r) as PermissionSetLicense;

if (!psl) throw new SfdxError('psl not found');
```

Reply:
@mshanemc I was mistaken regarding OOB error on an empty array.

Just tried `console.log([][0]);` and it yields `undefined`; Still thinking in Java.

Reply:
I'm going to put a modified version of this in core/Connection since that's already got the autoFetch query stuff.  Working on that as separate PR, then will put it back here maybe elsewhere.

I don't want lots of plugin commands to go through the same work of handling 0/1/1+ records

Reply:
@mshanemc, what relief are you proposing in core/Connection?

Reply:
`public async singleRecordQuery<T>(soql: string, options: SingleRecordQueryOptions = {}): Promise<T>`

* Executes a query using either standard REST or Tooling API, returning a single record.
* Will throw helpful error if either zero records are found OR multiple records are found.

Reply:
https://github.com/forcedotcom/sfdx-core/pull/350